# <center>  ETL </br>
#  <center> -- Extraer, Transformar y Cargar --</center>
Se realiza la extracción, transformación y carga de los datos proporcionados en el archivo 'output_steam_games.json', una primera aproximación a los datos con lo que se cuenta.


### <center>Importar Librerias</center>

In [67]:
import sys
sys.path.append("PROYECTO_INTEGRADOR_I")

In [68]:
import pandas as pd
import numpy as np
import json
import funciones
import os

***

### <center>Extracción de datos</center>

Se observa que el archivo es de tipo JSON, por lo cual se elige una forma particular para poder abrirlo. Se extraen los datos a un DataFrame.

In [69]:
#Se pasa dirección del archivo json
archivo = os.path.join('..', 'BD', 'output_steam_games.json')

# Se lee cada línea del dataset
registros= []
with open(archivo) as f:
    for line in f.readlines():
        data = json.loads(line)
        registros.append(data)

# Se transfoma en DataFrame
df_game = pd.DataFrame(registros)

***

In [70]:
#Visualización del DataFrame
df_game

publisher                                 genres  \
0                   NaN                                    NaN   
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title  \
0                            NaN                       NaN   
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date  \
0                                                     NaN          NaN   
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price early_access  \
0                                        

In [71]:
#(Filas, Columnas)
df_game.shape

(120445, 13)

In [72]:
#Nombre de las columnas
df_game.columns

Index(['publisher', 'genres', 'app_name', 'title', 'url', 'release_date',
       'tags', 'reviews_url', 'specs', 'price', 'early_access', 'id',
       'developer'],
      dtype='object')

### <center>Verificación de tipo de datos</center>

In [73]:
#Información de tipo de dato y nulos
df_game.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120445 entries, 0 to 120444
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher     24083 non-null  object
 1   genres        28852 non-null  object
 2   app_name      32133 non-null  object
 3   title         30085 non-null  object
 4   url           32135 non-null  object
 5   release_date  30068 non-null  object
 6   tags          31972 non-null  object
 7   reviews_url   32133 non-null  object
 8   specs         31465 non-null  object
 9   price         30758 non-null  object
 10  early_access  32135 non-null  object
 11  id            32133 non-null  object
 12  developer     28836 non-null  object
dtypes: object(13)
memory usage: 11.9+ MB


In [105]:
# Visualización más detallada de los tipos de datos, nulos y porcentajes por columna
funciones.verificar_tipo_datos(df_game)

Columna                              Tipo  NO_nulos_%  Nulos_%  Nulos
0        genres                   [<class 'str'>]      100.00     0.00      0
1         title  [<class 'str'>, <class 'float'>]      100.00     0.00      2
2   reviews_url                   [<class 'str'>]      100.00     0.00      0
3         price  [<class 'float'>, <class 'str'>]       95.53     4.47   3199
4       item_id                   [<class 'str'>]      100.00     0.00      0
5     developer  [<class 'str'>, <class 'float'>]       99.51     0.49    350
6  release_year                   [<class 'str'>]      100.00     0.00      0

Se observa valores nulos en algunos registros

***

### <center>Verificación de registros nulos </center>

In [76]:
# Visualización de las filas con todos los campos NaN
filas_nan = df_game[df_game.isnull().all(axis=1)]
filas_nan

publisher genres app_name title  url release_date tags reviews_url  \
0           NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
1           NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
2           NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
3           NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
4           NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
...         ...    ...      ...   ...  ...          ...  ...         ...   
88305       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
88306       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
88307       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
88308       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   
88309       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   

      specs price early_access   id developer  
0       NaN   NaN          NaN  NaN       NaN  
1       NaN   NaN          NaN  NaN       NaN  
2       NaN   NaN          NaN  NaN       NaN  
3       NaN   NaN          NaN  NaN       NaN  
4       NaN   NaN          NaN  NaN       NaN  
...     ...   ...          ...  ...       ...  
88305   NaN   NaN          NaN  NaN       NaN  
88306   NaN   NaN          NaN  NaN       NaN  
88307   NaN   NaN          NaN  NaN       NaN  
88308   NaN   NaN          NaN  NaN       NaN  
88309   NaN   NaN          NaN  NaN       NaN  

[88310 rows x 13 columns]

In [77]:
#Se observa 88310 filas con todo Nan y se decide eliminar dichos registros.
df_game = df_game.dropna(how='all').reset_index(drop=True)
df_game.shape

(32135, 13)

_____________

### <center>Verificación de registros duplicados </center>

In [78]:
#Visualización de registros idem
funciones.mostrar_filas_duplicadas(df_game, 'id')

publisher               genres  \
13894                                 Bethesda Softworks             [Action]   
14573                                 Bethesda Softworks             [Action]   
74                                                   NaN                  NaN   
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
13894                Wolfenstein II: The New Colossus   
14573                Wolfenstein II: The New Colossus   
74                                                NaN   
30961  Batman: Arkham City - Game of the Year Edition   

                                                title  \
13894                Wolfenstein II: The New Colossus   
14573                Wolfenstein II: The New Colossus   
74                                                NaN   
30961  Batman: Arkham City - Game of the Year Edition   

                                                     url release_date  \
13894          http://store.steampowered.com/app/612880/   2017-10-26   
14573  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26   
74                        http://store.steampowered.com/          NaN   
30961           http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags  \
13894  [Action, FPS, Gore, Violent, Alternate History...   
14573  [Action, FPS, Gore, Violent, Alternate History...   
74                                                   NaN   
30961  [Action, Open World, Batman, Adventure, Stealt...   

                                             reviews_url  \
13894  http://steamcommunity.com/app/612880/reviews/?...   
14573  http://steamcommunity.com/app/612880/reviews/?...   
74                                                   NaN   
30961                                                NaN   

                                                   specs  price early_access  \
13894  [Single-player, Steam Achievements, Full contr...  59.99        False   
14573  [Single-player, Steam Achievements, Full contr...  59.99        False   
74                                                   NaN  19.99        False   
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

           id                                   developer  
13894  612880                               Machine Games  
14573  612880                               Machine Games  
74        NaN                                         NaN  
30961     NaN  Rocksteady Studios,Feral Interactive (Mac)

Se observa que hay dos registros que son iguales, se procede a eliminar uno (13894 y 14573) y el 74 tiene valores nulos y la información que posee es insignificante, se elimina.

In [79]:
# se eliminan los index
df_game.drop(13894, inplace=True)
df_game.drop(74, inplace=True)

In [80]:
#se comprueba la eliminación registro 13894
if 13894 not in df_game.index:
    print("\nEl registro con índice 13894 se ha eliminado correctamente.")
else:
    print("\nEl registro con índice 13894 NO se ha eliminado correctamente.")


El registro con índice 13894 se ha eliminado correctamente.


In [81]:
#se comprueba la eliminación registro 74
if 74 not in df_game.index:
    print("\nEl registro con índice 74 se ha eliminado correctamente.")
else:
    print("\nEl registro con índice 74 NO se ha eliminado correctamente.")


El registro con índice 74 se ha eliminado correctamente.


In [82]:
#Se verificaque queda un ID = NAN
funciones.verificar_tipo_datos(df_game)

Columna                               Tipo  NO_nulos_%  Nulos_%  \
0      publisher   [<class 'str'>, <class 'float'>]       74.94    25.06   
1         genres  [<class 'list'>, <class 'float'>]       89.79    10.21   
2       app_name   [<class 'str'>, <class 'float'>]      100.00     0.00   
3          title   [<class 'str'>, <class 'float'>]       93.62     6.38   
4            url                    [<class 'str'>]      100.00     0.00   
5   release_date   [<class 'str'>, <class 'float'>]       93.57     6.43   
6           tags  [<class 'list'>, <class 'float'>]       99.50     0.50   
7    reviews_url   [<class 'str'>, <class 'float'>]      100.00     0.00   
8          specs  [<class 'list'>, <class 'float'>]       97.92     2.08   
9          price   [<class 'float'>, <class 'str'>]       95.71     4.29   
10  early_access                   [<class 'bool'>]      100.00     0.00   
11            id   [<class 'str'>, <class 'float'>]      100.00     0.00   
12     developer   [<class 'str'>, <class 'float'>]       89.74    10.26   

    Nulos  
0    8051  
1    3282  
2       1  
3    2049  
4       0  
5    2066  
6     162  
7       1  
8     669  
9    1377  
10      0  
11      1  
12   3298

In [83]:
# Se filtra las filas con 'id' nulos
filas_id_nulos = df_game[df_game['id'].isnull()]

# Se muestra el resultado
filas_id_nulos

publisher               genres  \
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
30961  Batman: Arkham City - Game of the Year Edition   

                                                title  \
30961  Batman: Arkham City - Game of the Year Edition   

                                            url release_date  \
30961  http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags reviews_url  \
30961  [Action, Open World, Batman, Adventure, Stealt...         NaN   

                                                   specs  price early_access  \
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

        id                                   developer  
30961  NaN  Rocksteady Studios,Feral Interactive (Mac)

Al observar que el id es NAN, buscamos por titulo = "Batman: Arkham City - Game of the Year Edition"

In [84]:
#Se busca según el título
df_game[df_game['title']=='Batman: Arkham City - Game of the Year Edition']

publisher               genres  \
1068   Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
1068   Batman: Arkham City - Game of the Year Edition   
30961  Batman: Arkham City - Game of the Year Edition   

                                                title  \
1068   Batman: Arkham City - Game of the Year Edition   
30961  Batman: Arkham City - Game of the Year Edition   

                                                     url release_date  \
1068   http://store.steampowered.com/app/200260/Batma...   2012-09-07   
30961           http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags  \
1068   [Action, Open World, Batman, Adventure, Stealt...   
30961  [Action, Open World, Batman, Adventure, Stealt...   

                                             reviews_url  \
1068   http://steamcommunity.com/app/200260/reviews/?...   
30961                                                NaN   

                                                   specs  price early_access  \
1068   [Single-player, Steam Achievements, Steam Trad...  19.99        False   
30961  [Single-player, Steam Achievements, Steam Trad...  19.99        False   

           id                                   developer  
1068   200260  Rocksteady Studios,Feral Interactive (Mac)  
30961     NaN  Rocksteady Studios,Feral Interactive (Mac)

In [85]:
#se elimina al comprobar que es igual
df_game.drop(30961, inplace=True)

In [86]:
#Se verifica que no queden nulos en ID
funciones.verificar_tipo_datos(df_game)

Columna                               Tipo  NO_nulos_%  Nulos_%  \
0      publisher   [<class 'str'>, <class 'float'>]       74.94    25.06   
1         genres  [<class 'list'>, <class 'float'>]       89.79    10.21   
2       app_name   [<class 'str'>, <class 'float'>]      100.00     0.00   
3          title   [<class 'str'>, <class 'float'>]       93.62     6.38   
4            url                    [<class 'str'>]      100.00     0.00   
5   release_date   [<class 'str'>, <class 'float'>]       93.57     6.43   
6           tags  [<class 'list'>, <class 'float'>]       99.50     0.50   
7    reviews_url                    [<class 'str'>]      100.00     0.00   
8          specs  [<class 'list'>, <class 'float'>]       97.92     2.08   
9          price   [<class 'float'>, <class 'str'>]       95.71     4.29   
10  early_access                   [<class 'bool'>]      100.00     0.00   
11            id                    [<class 'str'>]      100.00     0.00   
12     developer   [<class 'str'>, <class 'float'>]       89.74    10.26   

    Nulos  
0    8051  
1    3282  
2       1  
3    2049  
4       0  
5    2066  
6     162  
7       0  
8     669  
9    1377  
10      0  
11      0  
12   3298

__________________________________

### <center>Extraer AÑO de lanzamiento del item </center>
### <center>release_date </center>

Se debe crear una nueva columna release_year, que solo contiene el año, y se elimina release_date

In [87]:
# Se crea columna nueva con el año 'release_year'
df_game['release_year'] = df_game['release_date'].apply(funciones.obtener_release_year)

# Se elimina la columna 'release_date'
df_game = df_game.drop('release_date', axis=1)
df_game.head()

publisher                                             genres  \
0         Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1  Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2      Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3              彼岸领域                        [Action, Adventure, Casual]   
4               NaN                                                NaN   

                  app_name                    title  \
0      Lost Summoner Kitty      Lost Summoner Kitty   
1                Ironbound                Ironbound   
2  Real Pool 3D - Poolians  Real Pool 3D - Poolians   
3                  弹炸人2222                  弹炸人2222   
4            Log Challenge                      NaN   

                                                 url  \
0  http://store.steampowered.com/app/761140/Lost_...   
1  http://store.steampowered.com/app/643980/Ironb...   
2  http://store.steampowered.com/app/670290/Real_...   
3     http://store.steampowered.com/app/767400/2222/   
4  http://store.steampowered.com/app/773570/Log_C...   

                                                tags  \
0      [Strategy, Action, Indie, Casual, Simulation]   
1  [Free to Play, Strategy, Indie, RPG, Card Game...   
2  [Free to Play, Simulation, Sports, Casual, Ind...   
3                        [Action, Adventure, Casual]   
4                    [Action, Indie, Casual, Sports]   

                                         reviews_url  \
0  http://steamcommunity.com/app/761140/reviews/?...   
1  http://steamcommunity.com/app/643980/reviews/?...   
2  http://steamcommunity.com/app/670290/reviews/?...   
3  http://steamcommunity.com/app/767400/reviews/?...   
4  http://steamcommunity.com/app/773570/reviews/?...   

                                               specs         price  \
0                                    [Single-player]          4.99   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
2  [Single-player, Multi-player, Online Multi-Pla...  Free to Play   
3                                    [Single-player]          0.99   
4  [Single-player, Full controller support, HTC V...          2.99   

  early_access      id         developer        release_year  
0        False  761140         Kotoshiro                2018  
1        False  643980  Secret Level SRL                2018  
2        False  670290      Poolians.com                2017  
3        False  767400              彼岸领域                2017  
4        False  773570               NaN  Dato no disponible

In [88]:
# Se filtra las filas con 'Dato no disponible' en 'release_year'
registros_no_disponibles = df_game.loc[df_game['release_year'] == 'Dato no disponible']
registros_no_disponibles

publisher  \
4                                 NaN   
10                            Qucheza   
11                                NaN   
19                                NaN   
20                                NaN   
...                               ...   
32087  Greyson Richey, Nicholas Lives   
32088                rechargecomplete   
32097             Lunar Giant Studios   
32123                         VersoVR   
32134                             NaN   

                                                  genres  \
4                                                    NaN   
10             [Action, Indie, Simulation, Early Access]   
11                                                   NaN   
19                                                   NaN   
20                                                   NaN   
...                                                  ...   
32087                [Action, Casual, Indie, Simulation]   
32088           [Action, Adventure, Indie, Early Access]   
32097                                    [Action, Indie]   
32123  [Casual, Indie, Simulation, Sports, Early Access]   
32134                                                NaN   

                   app_name              title  \
4             Log Challenge                NaN   
10          Uncanny Islands    Uncanny Islands   
11     Icarus Six Sixty Six                NaN   
19            After Life VR                NaN   
20               Kitty Hawk                NaN   
...                     ...                ...   
32087      Lonely Astronaut   Lonely Astronaut   
32088     RECHARGE COMPLETE  RECHARGE COMPLETE   
32097            Tetrapulse         Tetrapulse   
32123          Cricket Club       Cricket Club   
32134           Maze Run VR                NaN   

                                                     url  \
4      http://store.steampowered.com/app/773570/Log_C...   
10     http://store.steampowered.com/app/768570/Uncan...   
11     http://store.steampowered.com/app/724910/Icaru...   
19     http://store.steampowered.com/app/772590/After...   
20     http://store.steampowered.com/app/640250/Kitty...   
...                                                  ...   
32087  http://store.steampowered.com/app/755830/Lonel...   
32088  http://store.steampowered.com/app/708070/RECHA...   
32097  http://store.steampowered.com/app/250440/Tetra...   
32123  http://store.steampowered.com/app/772180/Crick...   
32134  http://store.steampowered.com/app/681550/Maze_...   

                                                    tags  \
4                        [Action, Indie, Casual, Sports]   
10     [Early Access, Action, Indie, Simulation, Surv...   
11                                              [Casual]   
19                             [Early Access, Indie, VR]   
20      [Early Access, Action, Adventure, Indie, Casual]   
...                                                  ...   
32087                [Casual, Action, Indie, Simulation]   
32088           [Early Access, Action, Adventure, Indie]   
32097                             [Indie, Action, Co-op]   
32123  [Early Access, Indie, Casual, Simulation, Sports]   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  \
4      http://steamcommunity.com/app/773570/reviews/?...   
10     http://steamcommunity.com/app/768570/reviews/?...   
11     http://steamcommunity.com/app/724910/reviews/?...   
19     http://steamcommunity.com/app/772590/reviews/?...   
20     http://steamcommunity.com/app/640250/reviews/?...   
...                                                  ...   
32087  http://steamcommunity.com/app/755830/reviews/?...   
32088  http://steamcommunity.com/app/708070/reviews/?...   
32097  http://steamcommunity.com/app/250440/reviews/?...   
32123  http://steamcommunity.com/app/772180/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                  

In [89]:
#Se verifica que no queden nulos en release_year
funciones.verificar_tipo_datos(df_game)

Columna                               Tipo  NO_nulos_%  Nulos_%  \
0      publisher   [<class 'str'>, <class 'float'>]       74.94    25.06   
1         genres  [<class 'list'>, <class 'float'>]       89.79    10.21   
2       app_name   [<class 'str'>, <class 'float'>]      100.00     0.00   
3          title   [<class 'str'>, <class 'float'>]       93.62     6.38   
4            url                    [<class 'str'>]      100.00     0.00   
5           tags  [<class 'list'>, <class 'float'>]       99.50     0.50   
6    reviews_url                    [<class 'str'>]      100.00     0.00   
7          specs  [<class 'list'>, <class 'float'>]       97.92     2.08   
8          price   [<class 'float'>, <class 'str'>]       95.71     4.29   
9   early_access                   [<class 'bool'>]      100.00     0.00   
10            id                    [<class 'str'>]      100.00     0.00   
11     developer   [<class 'str'>, <class 'float'>]       89.74    10.26   
12  release_year                    [<class 'str'>]      100.00     0.00   

    Nulos  
0    8051  
1    3282  
2       1  
3    2049  
4       0  
5     162  
6       0  
7     669  
8    1377  
9       0  
10      0  
11   3298  
12      0

In [90]:
# Se obtiene la cantidad de apariciones de cada valor en 'release_year'
conteo_valores = df_game['release_year'].value_counts()

# Se muestra los valores y sus conteos
print(conteo_valores)

release_year
2017                  9517
2016                  6933
2015                  4945
2014                  2832
Dato no disponible    2351
2013                  1551
2012                  1191
2011                   589
2010                   436
2009                   337
2008                   213
2007                   167
2006                   155
2005                   102
2018                    96
2003                    84
2004                    79
2001                    69
1998                    61
1999                    55
2002                    50
1997                    42
2000                    40
1996                    39
1994                    36
1993                    34
1995                    33
1991                    16
1992                    15
1990                    12
1989                    10
1987                     9
1988                     7
1984                     4
1983                     4
1985                     3
1982           

_______________________________


### <center>Columna 'genres' </center>
<center>es una lista, hay que separarla</center>

In [91]:
# Se visualiza el dataframe
df_game

publisher                                             genres  \
0             Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1      Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2          Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3                  彼岸领域                        [Action, Adventure, Casual]   
4                   NaN                                                NaN   
...                 ...                                                ...   
32130   Ghost_RUS Games              [Casual, Indie, Simulation, Strategy]   
32131            Sacada                          [Casual, Indie, Strategy]   
32132      Laush Studio                        [Indie, Racing, Simulation]   
32133          SIXNAILS                                    [Casual, Indie]   
32134               NaN                                                NaN   

                       app_name                     title  \
0           Lost Summoner Kitty       Lost Summoner Kitty   
1                     Ironbound                 Ironbound   
2       Real Pool 3D - Poolians   Real Pool 3D - Poolians   
3                       弹炸人2222                   弹炸人2222   
4                 Log Challenge                       NaN   
...                         ...                       ...   
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url  \
0      http://store.steampowered.com/app/761140/Lost_...   
1      http://store.steampowered.com/app/643980/Ironb...   
2      http://store.steampowered.com/app/670290/Real_...   
3         http://store.steampowered.com/app/767400/2222/   
4      http://store.steampowered.com/app/773570/Log_C...   
...                                                  ...   
32130  http://store.steampowered.com/app/773640/Colon...   
32131  http://store.steampowered.com/app/733530/LOGis...   
32132  http://store.steampowered.com/app/610660/Russi...   
32133  http://store.steampowered.com/app/658870/EXIT_...   
32134  http://store.steampowered.com/app/681550/Maze_...   

                                                    tags  \
0          [Strategy, Action, Indie, Casual, Simulation]   
1      [Free to Play, Strategy, Indie, RPG, Card Game...   
2      [Free to Play, Simulation, Sports, Casual, Ind...   
3                            [Action, Adventure, Casual]   
4                        [Action, Indie, Casual, Sports]   
...                                                  ...   
32130              [Strategy, Indie, Casual, Simulation]   
32131                          [Strategy, Indie, Casual]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32134  [Early Access, Adventure, Indie, Action, Simul...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
1      http://steamcommunity.com/app/643980/reviews/?...   
2      http://steamcommunity.com/app/670290/reviews/?...   
3      http://steamcommunity.com/app/767400/reviews/?...   
4      http://steamcommunity.com/app/773570/reviews/?...   
...                                                  ...   
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs         price  \
0                                        [Single-player]          4.99   
1      [Single-player, Multi-play

In [92]:
# Se utilizar explode en la columna 'genres' para transformar las listas en filas duplicadas, manteniendo las demás columnas asociadas. 
df_game = df_game.explode('genres')

In [93]:
# Se verifica que no queden nulos en genres
funciones.verificar_tipo_datos(df_game)

Columna                               Tipo  NO_nulos_%  Nulos_%  \
0      publisher   [<class 'str'>, <class 'float'>]       79.16    20.84   
1         genres   [<class 'str'>, <class 'float'>]       95.61     4.39   
2       app_name   [<class 'str'>, <class 'float'>]      100.00     0.00   
3          title   [<class 'str'>, <class 'float'>]       97.26     2.74   
4            url                    [<class 'str'>]      100.00     0.00   
5           tags  [<class 'list'>, <class 'float'>]       99.75     0.25   
6    reviews_url                    [<class 'str'>]      100.00     0.00   
7          specs  [<class 'list'>, <class 'float'>]       98.74     1.26   
8          price   [<class 'float'>, <class 'str'>]       95.53     4.47   
9   early_access                   [<class 'bool'>]      100.00     0.00   
10            id                    [<class 'str'>]      100.00     0.00   
11     developer   [<class 'str'>, <class 'float'>]       95.35     4.65   
12  release_year                    [<class 'str'>]      100.00     0.00   

    Nulos  
0   15593  
1    3282  
2       2  
3    2050  
4       0  
5     184  
6       0  
7     940  
8    3346  
9       0  
10      0  
11   3478  
12      0

In [94]:
# Se eliminan los nulos en genero
df_game = df_game.dropna(subset=['genres'])

In [95]:
# Se visualizan nuevamente los datos en el dataframe
df_game

publisher      genres             app_name                title  \
0         Kotoshiro      Action  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro      Casual  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro       Indie  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro  Simulation  Lost Summoner Kitty  Lost Summoner Kitty   
0         Kotoshiro    Strategy  Lost Summoner Kitty  Lost Summoner Kitty   
...             ...         ...                  ...                  ...   
32132  Laush Studio       Indie        Russian Roads        Russian Roads   
32132  Laush Studio      Racing        Russian Roads        Russian Roads   
32132  Laush Studio  Simulation        Russian Roads        Russian Roads   
32133      SIXNAILS      Casual  EXIT 2 - Directions  EXIT 2 - Directions   
32133      SIXNAILS       Indie  EXIT 2 - Directions  EXIT 2 - Directions   

                                                     url  \
0      http://store.steampowered.com/app/761140/Lost_...   
0      http://store.steampowered.com/app/761140/Lost_...   
0      http://store.steampowered.com/app/761140/Lost_...   
0      http://store.steampowered.com/app/761140/Lost_...   
0      http://store.steampowered.com/app/761140/Lost_...   
...                                                  ...   
32132  http://store.steampowered.com/app/610660/Russi...   
32132  http://store.steampowered.com/app/610660/Russi...   
32132  http://store.steampowered.com/app/610660/Russi...   
32133  http://store.steampowered.com/app/658870/EXIT_...   
32133  http://store.steampowered.com/app/658870/EXIT_...   

                                                    tags  \
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
0          [Strategy, Action, Indie, Casual, Simulation]   
...                                                  ...   
32132                        [Indie, Simulation, Racing]   
32132                        [Indie, Simulation, Racing]   
32132                        [Indie, Simulation, Racing]   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   

                                             reviews_url  \
0      http://steamcommunity.com/app/761140/reviews/?...   
0      http://steamcommunity.com/app/761140/reviews/?...   
0      http://steamcommunity.com/app/761140/reviews/?...   
0      http://steamcommunity.com/app/761140/reviews/?...   
0      http://steamcommunity.com/app/761140/reviews/?...   
...                                                  ...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   

                                                   specs price early_access  \
0                                        [Single-player]  4.99        False   
0                                        [Single-player]  4.99        False   
0                                        [Single-player]  4.99        False   
0                                        [Single-player]  4.99        False   
0                                        [Single-player]  4.99        False   
...                                                  ...   ...          ...   
32132  [Single-player, Steam Achievements, Steam Trad...  1.99        False   
32132  [Single-player, Steam Achievements, Steam Trad...  1.99        False   
32132  [Single-player, Steam Achievements, Steam Trad...  1.99        False   
32133   [Single-player, Steam Achievements, Steam Cloud]  4.99        False   
32133   [Single-player, Steam Achievements, Steam Clo

____________________

Se procede a eliminar las columnas consideradas no necesarias para el estudio a realizar

In [96]:
# Se elimina las columnas que no necesitamos del Dataframe
df_game = df_game.drop('publisher', axis=1)
df_game = df_game.drop('url', axis=1)
df_game = df_game.drop('app_name', axis=1)
df_game = df_game.drop('specs', axis=1)
df_game = df_game.drop('early_access', axis=1)
df_game = df_game.drop('tags', axis=1)

In [97]:
# Se renombra la columna "id" por "item_id" para su posterior relación con otras tablas
df_game.rename(columns={'id': 'item_id'}, inplace=True)

In [100]:
# Se visualiza el dataframe limpio
df_game

genres                title  \
0          Action  Lost Summoner Kitty   
0          Casual  Lost Summoner Kitty   
0           Indie  Lost Summoner Kitty   
0      Simulation  Lost Summoner Kitty   
0        Strategy  Lost Summoner Kitty   
...           ...                  ...   
32132       Indie        Russian Roads   
32132      Racing        Russian Roads   
32132  Simulation        Russian Roads   
32133      Casual  EXIT 2 - Directions   
32133       Indie  EXIT 2 - Directions   

                                             reviews_url price item_id  \
0      http://steamcommunity.com/app/761140/reviews/?...  4.99  761140   
0      http://steamcommunity.com/app/761140/reviews/?...  4.99  761140   
0      http://steamcommunity.com/app/761140/reviews/?...  4.99  761140   
0      http://steamcommunity.com/app/761140/reviews/?...  4.99  761140   
0      http://steamcommunity.com/app/761140/reviews/?...  4.99  761140   
...                                                  ...   ...     ...   
32132  http://steamcommunity.com/app/610660/reviews/?...  1.99  610660   
32132  http://steamcommunity.com/app/610660/reviews/?...  1.99  610660   
32132  http://steamcommunity.com/app/610660/reviews/?...  1.99  610660   
32133  http://steamcommunity.com/app/658870/reviews/?...  4.99  658870   
32133  http://steamcommunity.com/app/658870/reviews/?...  4.99  658870   

                      developer release_year  
0                     Kotoshiro         2018  
0                     Kotoshiro         2018  
0                     Kotoshiro         2018  
0                     Kotoshiro         2018  
0                     Kotoshiro         2018  
...                         ...          ...  
32132  Laush Dmitriy Sergeevich         2018  
32132  Laush Dmitriy Sergeevich         2018  
32132  Laush Dmitriy Sergeevich         2018  
32133             xropi,stev3ns         2017  
32133             xropi,stev3ns         2017  

[71551 rows x 7 columns]

In [98]:
# Se verifica los tipos de datos y nulos
funciones.verificar_tipo_datos(df_game)

Columna                              Tipo  NO_nulos_%  Nulos_%  Nulos
0        genres                   [<class 'str'>]      100.00     0.00      0
1         title  [<class 'str'>, <class 'float'>]      100.00     0.00      2
2   reviews_url                   [<class 'str'>]      100.00     0.00      0
3         price  [<class 'float'>, <class 'str'>]       95.53     4.47   3199
4       item_id                   [<class 'str'>]      100.00     0.00      0
5     developer  [<class 'str'>, <class 'float'>]       99.51     0.49    350
6  release_year                   [<class 'str'>]      100.00     0.00      0

In [107]:
# Se contabiliza la aparición de cada valor que toma la columna 'price'
conteo_publisher = df_game['price'].value_counts()
conteo_publisher

price
4.99                         9475
9.99                         8692
2.99                         7583
0.99                         6190
1.99                         5623
                             ... 
27.49                           1
42.99                           1
14.95                           1
0.5                             1
Free HITMAN™ Holiday Pack       1
Name: count, Length: 150, dtype: int64

In [109]:
# Se reemplaza 'Free To Play' por 0 y se convierte la columna 'price' a tipo float
df_game['price'] = pd.to_numeric(df_game['price'].replace('Free To Play', 0), errors='coerce')

# Reemplazar NaN por 0
df_game['price'] = df_game['price'].fillna(0)

In [110]:
funciones.verificar_tipo_datos(df_game)

Columna                              Tipo  NO_nulos_%  Nulos_%  Nulos
0        genres                   [<class 'str'>]      100.00     0.00      0
1         title  [<class 'str'>, <class 'float'>]      100.00     0.00      2
2   reviews_url                   [<class 'str'>]      100.00     0.00      0
3         price                 [<class 'float'>]      100.00     0.00      0
4       item_id                   [<class 'str'>]      100.00     0.00      0
5     developer  [<class 'str'>, <class 'float'>]       99.51     0.49    350
6  release_year                   [<class 'str'>]      100.00     0.00      0

________________

### <center> terminada la limpieza se guarda en un archivo csv </center>

In [111]:
# Se construye la ruta completa al archivo CSV desde la ubicación del notebook
game = os.path.join('..', 'CSV', 'games_limpio.csv')

# Se guarda el DataFrame en el archivo CSV
df_game.to_csv(game, index=False, encoding='utf-8')

# Se imprime un mensaje si se guardó correctamente
print(f'Se guardó el archivo {game}')

Se guardó el archivo ..\CSV\games_limpio.csv
